From 396fd136630876e4a43e4a55ee161779bc224e68 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Quick=20=E2=9A=A1?= <quick@disroot.org>
Date: Sat, 8 Nov 2025 14:59:47 +0530
Subject: [PATCH 6/6] Settings: Allow disabling ReVanced mods [2/2]

* http://github.com/PixelLineage/vendor_revanced
* Uninstall updates upon toggle to avoid issues with stock app

Change-Id: I40b593b1b16ef687086e6b5a1586bfe22057fa16
---
 res/values/custom_strings.xml                 |  6 ++
 res/xml/apps.xml                              |  6 ++
 .../applications/AppDashboardFragment.java    | 72 +++++++++++++++++++
 3 files changed, 84 insertions(+)

diff --git a/res/values/custom_strings.xml b/res/values/custom_strings.xml
index f53367b1788..4088340d515 100644
--- a/res/values/custom_strings.xml
+++ b/res/values/custom_strings.xml
@@ -384,4 +384,10 @@
     <string name="hide_developer_status_title">Hide developer status</string>
     <string name="hide_developer_status_summary">Hide developer status from apps</string>
 
+    <!-- ReVanced -->
+    <string name="revan_toggle_title">ReVanced Integration</string>
+    <string name="revan_toggle_summary">Use ReVanced variants of YouTube and YT Music instead of stock apps</string>
+    <string name="revan_restart_dialog_title">Restart device?</string>
+    <string name="revan_restart_dialog_message">Restart now to apply this change.</string>
+    <string name="revan_restart_now">Restart now</string>
 </resources>
diff --git a/res/xml/apps.xml b/res/xml/apps.xml
index ff69059b6bc..5d81f666cfb 100644
--- a/res/xml/apps.xml
+++ b/res/xml/apps.xml
@@ -75,6 +75,12 @@
             android:targetPackage="${applicationId}"/>
     </Preference>
 
+    <SwitchPreferenceCompat
+        android:key="persist_revan_mod"
+        android:title="@string/revan_toggle_title"
+        android:summary="@string/revan_toggle_summary"
+        android:defaultValue="true" />
+
     <PreferenceCategory
         android:key="dashboard_tile_placeholder"
         android:order="10"/>
diff --git a/src/com/android/settings/applications/AppDashboardFragment.java b/src/com/android/settings/applications/AppDashboardFragment.java
index 74b2f31fd97..ce2360c4be2 100644
--- a/src/com/android/settings/applications/AppDashboardFragment.java
+++ b/src/com/android/settings/applications/AppDashboardFragment.java
@@ -20,6 +20,10 @@ import android.app.Activity;
 import android.app.settings.SettingsEnums;
 import android.content.Context;
 import android.content.Intent;
+import android.content.pm.IPackageDeleteObserver;
+import android.content.pm.PackageManager;
+import android.os.PowerManager;
+import android.os.SystemProperties;
 import android.provider.SearchIndexableResource;
 import android.net.Uri;
 import android.os.Bundle;
@@ -31,6 +35,9 @@ import androidx.preference.Preference;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
+import androidx.appcompat.app.AlertDialog;
+import androidx.preference.Preference;
+import androidx.preference.SwitchPreferenceCompat;
 
 import com.android.internal.annotations.VisibleForTesting;
 import com.android.settings.R;
@@ -41,9 +48,12 @@ import com.android.settings.widget.PreferenceCategoryController;
 import com.android.settingslib.core.AbstractPreferenceController;
 import com.android.settingslib.search.SearchIndexable;
 
+import static android.os.UserHandle.USER_SYSTEM;
+
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
 
 /** Settings page for apps. */
 // LINT.IfChange
@@ -59,6 +69,12 @@ public class AppDashboardFragment extends DashboardFragment {
     private ActivityResultLauncher<Intent> mPifFilePickerLauncher;
     private KeyboxDataPreference mKeyboxDataPreference;
     private PifDataPreference mPifDataPreference;
+    private static final String REVAN_PREF_KEY = "persist_revan_mod";
+    private static final String REVAN_PROP = "persist.sys.revan.mod";
+    private static final String[] REVAN_PACKAGES = {
+            "com.google.android.youtube",
+            "com.google.android.apps.youtube.music"
+    };
     private AppsPreferenceController mAppsPreferenceController;
 
     private static List<AbstractPreferenceController> buildPreferenceControllers(Context context) {
@@ -153,6 +169,62 @@ public class AppDashboardFragment extends DashboardFragment {
         }
     }
 
+    @Override
+    public void onStart() {
+        super.onStart();
+
+        final Preference pref = findPreference(REVAN_PREF_KEY);
+        if (pref instanceof SwitchPreferenceCompat) {
+            final SwitchPreferenceCompat toggle = (SwitchPreferenceCompat) pref;
+            toggle.setChecked(SystemProperties.getBoolean(REVAN_PROP, true));
+            toggle.setOnPreferenceChangeListener((p, newVal) -> {
+                showRevanRestartDialog(toggle, (Boolean) newVal);
+                return false;
+            });
+        }
+    }
+
+    private void showRevanRestartDialog(SwitchPreferenceCompat toggle, boolean enabled) {
+        new AlertDialog.Builder(getContext())
+                .setTitle(R.string.revan_restart_dialog_title)
+                .setMessage(R.string.revan_restart_dialog_message)
+                .setPositiveButton(R.string.revan_restart_now, (dialog, which) -> {
+                    final Context context = requireContext().getApplicationContext();
+                    toggle.setChecked(enabled);
+                    if (enabled) {
+                        uninstallUpdatesThenEnable(context);
+                    } else {
+                        SystemProperties.set(REVAN_PROP, "false");
+                        reboot(context);
+                    }
+                })
+                .setNegativeButton(android.R.string.cancel, null)
+                .show();
+    }
+
+    private void uninstallUpdatesThenEnable(Context context) {
+        // Keep the property false while PackageManager restores stock system package paths.
+        final PackageManager pm = context.getPackageManager();
+        final AtomicInteger pendingDeletes = new AtomicInteger(REVAN_PACKAGES.length);
+        final IPackageDeleteObserver observer = new IPackageDeleteObserver.Stub() {
+            @Override
+            public void packageDeleted(String packageName, int returnCode) {
+                if (pendingDeletes.decrementAndGet() == 0) {
+                    SystemProperties.set(REVAN_PROP, "true");
+                    reboot(context);
+                }
+            }
+        };
+
+        for (String pkg : REVAN_PACKAGES) {
+            pm.deletePackageAsUser(pkg, observer, 0, USER_SYSTEM);
+        }
+    }
+
+    private void reboot(Context context) {
+        context.getSystemService(PowerManager.class).reboot(null);
+    }
+
     @VisibleForTesting
     PreferenceCategoryController getAdvancedAppsPreferenceCategoryController() {
         return use(AdvancedAppsPreferenceCategoryController.class);
-- 
2.34.1

